pkg/clustermesh/mcsapi/endpointslice_mirror_controller.go

```diff
@@ -0,0 +1,477 @@
+			switch obj.GetLabels()[discoveryv1.LabelManagedBy] {
+			case utils.EndpointSliceMeshControllerName:
+				// We can explicitly ignore remote EndpointSlice
+			case endpointSliceLocalMCSAPIControllerName:
+				return true
+			default:
+				return true
+			}
+
+			return false
+		}))).
+
+		// Special watchers for EndpointSlice deletion
+		Watches(&discoveryv1.EndpointSlice{}, &endpointSliceMirrorDeleteWatcher{}).
+
+		// Watch for changes to Service to enqueue local EndpointSlice
+		Watches(&mcsapiv1alpha1.ServiceExport{}, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []ctrl.Request {
+			serviceReq, _ := labels.NewRequirement(discoveryv1.LabelServiceName, selection.Equals, []string{obj.GetName()})
+			selector := labels.NewSelector()
+			selector = selector.Add(*serviceReq)
+
+			var epSliceList discoveryv1.EndpointSliceList
+			if err := r.Client.List(ctx, &epSliceList, &client.ListOptions{Namespace: obj.GetNamespace(), LabelSelector: selector}); err != nil {
+				return []ctrl.Request{}
+			}
+
+			requests := []ctrl.Request{}
+			for _, epSlice := range epSliceList.Items {
+				requests = append(requests, ctrl.Request{NamespacedName: types.NamespacedName{
+					Name:      epSlice.Name,
+					Namespace: epSlice.Namespace,
+				}})
+			}
+			return requests
+		})).
+		// Watch for changes to derived Service to enqueue local EndpointSlices.
+		// We need to enqueue the "other" EndpointSlice to allow derived
+		// EndpointSlice initial creation.
+		Watches(&corev1.Service{}, handler.EnqueueRequestsFromMapFunc(func(ctx context.Context, obj client.Object) []ctrl.Request {
+			svcImportOwner := getOwnerReferenceName(obj.GetOwnerReferences(), mcsapiv1alpha1.GroupVersion.String(), kindServiceImport)
+			if svcImportOwner != "" {
+				serviceReq, _ := labels.NewRequirement(discoveryv1.LabelServiceName, selection.Equals, []string{svcImportOwner})
+				selector := labels.NewSelector()
+				selector = selector.Add(*serviceReq)
+
+				var epSliceList discoveryv1.EndpointSliceList
+				if err := r.Client.List(ctx, &epSliceList, &client.ListOptions{Namespace: obj.GetNamespace(), LabelSelector: selector}); err != nil {
+					return []ctrl.Request{}
+				}
+
+				requests := []ctrl.Request{}
+				for _, epSlice := range epSliceList.Items {
+					requests = append(requests, ctrl.Request{NamespacedName: types.NamespacedName{
+						Name:      epSlice.Name,
+						Namespace: epSlice.Namespace,
+					}})
+				}
+				return requests
+			}
+			return []ctrl.Request{}
+		})).
+		Complete(r)
+}
+
+// getSuffix return the name of the EndpointSlice trimmed by its generatedName
+func getSuffix(endpointSlice *discoveryv1.EndpointSlice) string {
+	suffix := strings.TrimPrefix(endpointSlice.Name, endpointSlice.Labels[discoveryv1.LabelServiceName]+"-")
+	suffixLen := min(40, len(suffix))
+	suffix = strings.TrimPrefix(suffix[len(suffix)-suffixLen:], "-")
```

Comment:
Technically endpointslice names can also contain dots, which should be sanitized as well.

```suggestion
	suffix = strings.TrimLeft(suffix[len(suffix)-suffixLen:], "-.")
```